### DISCUSSION TITLE
Update returning orm object

### DISCUSSION THREAD

**Discussion by ave-satan** [sqlalchemy/sqlalchemy#8895] [Usage Questions] [answered]
I'm trying to issue an update query with `RETURNING` clause to get back the data with a single sql request.
```python
result = await session.execute(
    sa.update(SomeModel)
    .values(**data)
    .where(SomeModel.id == id)
    .returning(SomeModel)
)
```
The problem is that it returns columns tuple and not an actual orm object.
I got it working like this
```python
data = result.one()
some_model = SomeModel(**data)
```
But it looks kinda ugly. Is there a way to get an orm object at one blow without having to create it manually?

**Comment by zzzeek** [comment, marked as answer]
yes, for SQLAlchemy 1.4 see:

https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-insert-update-and-on-conflict-i-e-upsert-to-return-orm-objects

for SQLAlchemy 2.0 see:

https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-update-delete-and-custom-where-criteria

**Reply by ave-satan** [reply]
Thanks! I must have overlooked this when searching through the docs

**Reply by CaselIT** [reply]
@zzzeek on v2 does that update also require populate_existing or it's automatically used?

**Reply by zzzeek** [reply]
populate existing is not implicit, no.    eventually I want to have it so you can make a Session where you just turn on populate_existing for the whole session (constructor-level execution options, there's an issue for it).

**Reply by zzzeek** [reply]
> Thanks! I must have overlooked this when searching through the docs

the 1.4 version is quite buried because it's experimental.  2.0's docs feature these patterns much more prominently as a lot of fixes and development have been done.

**Reply by CaselIT** [reply]
> populate existing is not implicit, no. eventually I want to have it so you can make a Session where you just turn on populate_existing for the whole session (constructor-level execution options, there's an issue for it).

I didn't explain myself correctly, sorry.

What I meant is that the 1.4 docs use populate existing, while the 2.0 don't. But since update is now supported by the orm it probably triggers synchronize session, so the end result is the same for existing objects?

**Reply by zzzeek** [reply]
I knew what you meant, but as far as the answer, I'm not sure :)  OK, so if you are doing ORM enabled update/delete, then yes synchronize_session=auto is going to populate things.   Does having them in returning change that.    Um, I dont remember.  maybe.  soooo....we should make sure the docs mention what's expected here.

**Reply by CaselIT** [reply]
indeed or enable it by default since the session need to process the data anyway so it may as well update the objects

do you want to open an issue to track it?

**Reply by zzzeek** [reply]
I think it does this and just have to update the docs (or just read the docs, see if I noted it already)

...


https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-update-delete-and-custom-where-criteria

> The support for RETURNING is also compatible with the fetch synchronization strategy, which also uses RETURNING. The ORM will organize the columns in RETURNING appropriately so that the synchronization proceeds as well as that the returned [Result](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result) will contain the requested entities and SQL columns in their requested order.

that seems to confirm it.  then for upserts, we *dont* do the sync thing, they have to use populate existing, that's also noted:

https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#using-returning-with-upsert-statements

so we can see things are there, but we're going to still have a little confusion on this stuff.   The new DML features are a huge step forward though, I can see us in 2.1 making these two methods more consistent somehow.

**Reply by zzzeek** [reply]
like, what happens if I use update() with synchronize_session='fetch', *and* returning with populate existing.  I think that will work fine but is it doing redundant work in that case.   heh!

**Reply by CaselIT** [reply]
maybe for now we can just update the returning example to include populate existing so that the expected updated values are returned

**Reply by zzzeek** [reply]
ok....we havent seen anyone testing these features in the 2.0 betas so we didnt get any feedback yet
